### ISSUE TITLE
add password reovery and make signing up on the mailing list let you create an account

### ISSUE THREAD

**Issue by ospc-auto** [OpenSourcePolicyCenter/webapp-public#16]
_From @amfarrell on July 2, 2015 2:55_

There is no way for current users to recover passwords. Also, the current way that mailing-list sign-up is implemented (in the soon-to-be-merged mailchimp branch), it merely adds your email address to a database table. It does not create an account for you and thus does not let you generate budget projections. This is a proposal to fix both of these via the same mechanism.

[ ] Create a password recovery process for accounts.
  [ ] Create a Password recovery table consisting of the hash in the link sent to users, a time generated, and the ForeignKey to the users account. There is probably already a django plugin for this.
  [ ] Modify the sign-up page to include a "recover password" link.
  [ ] Create a landing page where users enter a new password. This page should not accept links that are more than a day old.
[ ] Use that same process for new account creation
  [ ] Create a new landing page where users enter a new password as well as other information we might want from them.

_Copied from original issue: OpenSourcePolicyCenter/webapp#223_

**Comment by ospc-auto** [comment]
_From @amfarrell on July 2, 2015 2:56_

@MattHJensen : Thoughts on this proposal?

**Comment by ospc-auto** [comment]
_From @MattHJensen on July 2, 2015 17:15_

Yes, @amfarrell, this makes sense. Let me add some background and a kink: 

**How we use accounts now**

We are currently waiting on the IRS to release a 2009 public use file. We are using the 2008 puf now. While we are waiting for the data update, the IRS has requested that we limit by hand who has access to TaxBrain. Currently I have been making accounts for select people and assigning them a password through the django admin. Essentially, if you have an account, you can use TaxBrain. 

**How we might use accounts in the future**

The status quo is an adequate solution for now, but not long-term sustainable. In the future, we will want all users to be able to use TaxBrain, whether they are logged in or not. Logging in might give you additional features (perhaps you might be able to store your simulated plans, with your name attached, in an OSPC tax-plan library), but it will also be important for many policy maker users that they can use TaxBrain completely anonymously, without logging in. 

It sounds like your proposal would allow for all of this. 

**The kink**

The other thing that having an account might allow you to do is receive special permissions for new features, say the dynamic model beta. 

**How this applies to today**
Under your proposal, I think we might also want to use this "special permissions" style for giving access to TaxBrain.

One way to do that might be to give a token to users that they could submit during the account creation process and which would give their account access to TaxBrain. Or alternatively, (perhaps to begin with) when the user creates an account, he would tell me his username and then I would add the permission through the admin. 

**Pitfalls**

A pitfall might be that if we are to make these changes, then a user would need to create a password to get on the mailing list. This could be frustrating for current users who don't have access to TaxBrain. It could be frustrating in the future for users who never want to log in to use TaxBrain. Is there a good way around this? Perhaps allow people to get on the mailing list with just an email address, and then signup with a password separately? I guess the problem would be if a user signs up for the mailing list with one email address and then makes an account with another -- they'd receive the newsletter to both emails (maybe not the worst thing).

**Summary**

My thinking on this is relatively fresh, so there might be completely different approach here.

**Comment by ospc-auto** [comment]
_From @zrisher on July 8, 2015 16:49_

A few thoughts here:

**Password recovery, new account creation page**
 It does indeed seem this is accomplished by existing code, in fact it's part of [`contrib.auth`](https://docs.djangoproject.com/en/1.8/topics/auth/default/#module-django.contrib.auth.views). It looks like we could do all of the recovery and replace some of our existing views by simply including those. I'm not sure how difficult the integration would be since we're not using the Django user models directly, but it seems worth considering. I would always try to use pre-existing code as much as possible, but you guys are the experts here and I wouldn't be surprised if you chose to go with custom user models and views because the default ones were lacking in some ways or hard to configure.

**Linking newsletter signups to accounts**
Agreed seems nice to have; it would make user activity analysis easier down the line and would allow a simple model for managing subscriptions. But it seems like a lot of work to allow accounts to exist without passwords being set on them, and it also seems like it's an important feature for users to be able to sign up for newsletters without entering any details besides their email. So I think the best link we might be able to achieve here is just looking up a user's subscriptions by querying the subscriptions table for their email.

**Comment by talumbau** [comment]
User login/authentication needs a re-think from top to bottom. Closing this so we can start fresh.
